### PR TITLE
ftrackComponents

### PR DESCRIPTION
**Pull request by tokejepsen**
Adding feature for "ftrackComponents, which is list of dictionaries with `value` and `name` for adding multiple components per instance.

Also slight rename of class to reflect file name.

### CODE REVIEW

File: pyblish_ftrack/plugins/conform_ftrack_component.py [hunk 1]
```diff
@@ -21,6 +21,34 @@ class FtrackUploadComponent(pyblish.api.Conformer):
...
+                    else:
+                        self.log.info('Component already exists.')
+
+                else:
+                    self.log.info('No versionID found in context.')
+        else:
+            self.log.info('No components found in instance.')
+
         if instance.has_data('ftrackComponent') and instance.has_data('ftrackComponentName'):
             sourcePath = os.path.normpath(instance.data('ftrackComponent'))
```
Comment (mottosso) [inline pyblish_ftrack/plugins/conform_ftrack_component.py:53]: This doesn't look right.

**Comment by mottosso** [comment]
How come you're showing up as `unknown`, @tokejepsen?

**Comment by tokejepsen** [comment]
I do't know, maybe something to do with my SourceTree settings.

**Comment by mottosso** [comment]
Should probably have a look at that, I think your email needs to match the one you're registered with on GitHub.

**Comment by tokejepsen** [comment]
Is it a problem for the pull-request?

**Comment by mottosso** [comment]
No, shouldn't be, it's just that you won't get credited for the commits you've made. :(

**Comment by tokejepsen** [comment]
Thats alright, I'll fix it at some point.

**Comment by mottosso** [comment]
Ok. Did you want @mkolar to have a look before I merge? I'm not the best judge as I can't test it.

In fact, would one of you like to have admin rights here and make decisions yourself? It probably makes more sense than me merging anything willy-nilly.

**Comment by mottosso** [comment]
But have a look at my code-comment, @tokejepsen, it looks like some left over code.

**Comment by tokejepsen** [comment]
> In fact, would one of you like to have admin rights here and make decisions yourself? It probably makes more sense than me merging anything willy-nilly.

Might be best to give @mkolar rights as he's the original author.

> But have a look at my code-comment

I don't quite get what I need to look at?

**Comment by mottosso** [comment]
I commented directly under the line, but as it's the final line it doesn't look very obvious it's true.

I'm referring to this.

``` python
    if instance.has_data('ftrackComponent') and instance.has_data('ftrackComponentName'):
             sourcePath = os.path.normpath(instance.data('ftrackComponent'))
```

Which assigns to sourcePath (which isn't snake_case, btw), but that variable isn't used. :S

**Comment by tokejepsen** [comment]
> Which assigns to sourcePath (which isn't snake_case, btw), but that variable isn't used. :S

Its part of the existing code, so sourcePath is used later in the plugin. Guess it just shows you what comes afterwards with a couple of lines.

**Comment by mottosso** [comment]
Ah, indeed you are right, did not expect that.

**Comment by mottosso** [comment]
Adding admin rights to Milan.

@mkolar, when you think the request is good, you should be able to merge it.

**Comment by mkolar** [comment]
A few things that I notice here. You should probably update docstring to reflect the changes. Ideally saying what type of data 'ftrackComponents' is, you only mentioned it is a commit comment. Then we should probably get rid of the original way of passing component as it's duplicate functionality and adds complexity for no reason. I reckon if user only want's to publish one component per instance (most cases I assume) he'll just add one dictionary in the ftrackComponents list. So lines 52 onwards should probably be removed.

Concerning the change of class name. I'd like some input from @mottosso here. I'm a bit sick of these awkward names 'ConformFtrackComponent', 'Conform.....' etc, appearing in the UI. I'm trying to make them more descriptive and clearer for artist, that's why I rather used 'UploadFtrackComponent'. We could call it something else like 'CreateFtrackComponent' or similar. I just don't see the point in adding Conform at the beginning of all of them. It's already in conform category so why stress it even more. It doesn't help the user in any way, just clutters the UI. It would make perfect sense to have class match the filename if there was a 'nice name' attribute that would show up in the UI. until then I'd rather stick with descriptive names when possible.

Other than that, it looks good.

**Comment by mottosso** [comment]
> Concerning the change of class name. I'd like some input from @mottosso here. I'm a bit sick of these awkward names 'ConformFtrackComponent', 'Conform.....' etc, appearing in the UI.

I agree. There's been [talk](https://github.com/pyblish/pyblish-qml/issues/85) about introducing an override (e.g. niceName) for plug-ins, and I think it's about time we get it in.

Stick with convention for now, and I'll push out a new version with the `name` property implemented by the end of next week.

**Comment by mottosso** [comment]
About the admin rights, @mkolar, one thing. Try not to push directly into the pyblish/pyblish-ftrack project. Always push into your fork and do a pull-request. It has some technical advantages but also social in that we can see and talk about changes here.

**Comment by mottosso** [comment]
About the PR, what do you guys think about these log messages?

``` python
# This looks to be for the artist
self.log.info('Component already exists.')
else:
  # But this and the next looks to be for the plug-in developer
  # What can the artist do about a missing versionID in this "context"?
  # They might not even know what either of those are.
  self.log.info('No versionID found in context.')
      else:
          self.log.info('No components found in instance.')
```

**Comment by tokejepsen** [comment]
> About the PR, what do you guys think about these log messages?

I was thinking about the log messaging as well, as its getting a bit cluttered atm. I know its a bit difficult to know what to show the artist, but maybe we should keep the logging to debugging state unless its really important for the user.

> Stick with convention for now, and I'll push out a new version with the name property implemented by the end of next week.

@mkolar we could make an issue for now about changing the names when the feature is implemented?

**Comment by mkolar** [comment]
> About the admin rights, @mkolar, one thing. Try not to push directly into the pyblish/pyblish-ftrack project. 

Sure. I believe that's what I've done so far. I'm just working with my fork.

> About the PR, what do you guys think about these log messages?

I think we should keep them there, but change the message for something more 'user friendly' 

e.g.
'No ftrack version found for components to be uploaded to.'
'No components can be created from this instance'

**Comment by tokejepsen** [comment]
> I think we should keep them there, but change the message for something more 'user friendly'

As it not critical and the publish can run through fine and still show some warning messages, maybe change those to info states?

**Comment by mottosso** [comment]
> Sure. I believe that's what I've done so far. I'm just working with my fork.

Yes, you have, it was meant only as a precaution for the future.

> @tokejepsen but maybe we should keep the logging to debugging state unless its really important for the user.
> @mkolar I think we should keep them there, but change the message for something more 'user friendly'

+1, you could even do both. Keep user-friendlyness to the `INFO` level, and messages more useful  to you at `DEBUG`. Debug should be hidden unless the user enables it, so you could chuck lots of info in there if it helps you debug. They are also part of the `result` objects found in `context.data("results")` so you could dig into them port-mortem if you store it somewhere, like in Firebase. @tokejepsen knows what I'm talking about.

**Comment by mkolar** [comment]
Very true. I'll do similar changes throughout the extension as well. To be honest I totally forgot about the `DEBUG` log.

**Comment by mottosso** [comment]
Don't forget about warning, error and critical either. They show up with increasingly vibrant colors in the terminal.

**Comment by mkolar** [comment]
warning I'm using. Error I havent' found use for yet, considering that if it happens I probably want to raise it as normal exception.

**Comment by tokejepsen** [comment]
closing this in favour of more recent pull-request

**Pull request closed without merging**